### PR TITLE
Support for FIPS-Approved Submodule

### PR DESCRIPTION
**Pull request by m4dc4p**
* Added "CipherSuite" enum to indicate which cipher suite the crypto library is using
* Updated Client to write/read Sodium (Curve25519/Ed25519) keys or FIPS keys (P384).
* Updated some tests that depended on Sodium features
* Conditional build for FIPS-approved mode available when e3db-fips module is present (private repo).
* Update Android "targetSdk" & compileSdk versions to 27; API 16 minimum still supported.
* Updated large file support to work around bug in LazySodium library; previous file versions still supported.

### CODE REVIEW

**Review by emanb29 — APPROVED** [review]
Great documentation, and very clear structure. All looks good to me.

File: e3db-crypto-interface/src/main/java/com/tozny/e3db/CipherSuite.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    this.encryptionKey = encryptionKey;
+    this.signingKey = signingKey;
+  }
+
+  public KeyType getSigningKey() {
+    return signingKey;
+  }
+
+  public KeyType getEncryptionKey() {
```
Comment (gstro) [inline e3db-crypto-interface/src/main/java/com/tozny/e3db/CipherSuite.java]: Could be good to have these say `get_KeyType()` so it doesn't look like they return bytes when called.

File: e3db/src/main/java/com/tozny/e3db/Client.java [hunk 2]
```diff
@@ -989,15 +1001,39 @@ private EAKEntry getEAK(UUID writerId, UUID userId, UUID readerId, String type)
...
+            case Curve25519:
+              publicKey = eakResponse.get("authorizer_public_key").get("curve25519").asText();
+              break;
+            case P384:
+              publicKey = eakResponse.get("authorizer_public_key").get("p384").asText();
+              break;
+            default:
+              throw new IllegalStateException("Encryption key type " + Platform.crypto.suite().getEncryptionKey() + " not supported.");
+          }
+        }
```
Comment (gstro) [inline e3db/src/main/java/com/tozny/e3db/Client.java]: Could be good to keep this DRY.

**Review by gstro — APPROVED** [review]
LGTM

File: e3db/src/main/java/com/tozny/e3db/Client.java (hunk 2 above)
Comment (m4dc4p) [reply e3db/src/main/java/com/tozny/e3db/Client.java]: I agree but I couldnt think of a way that didn't seem brittle.

File: androidtest/src/androidTest/java/com/tozny/e3db/AndroidCryptoTest.java [hunk 3]
```diff
@@ -360,4 +360,41 @@ public void testRoundtripFile() throws IOException {
...
+      int a, b, pos = 1;
+      for(a = expected.read(), b = actual.read(); a != -1 && b != -1; a = expected.read(), b = actual.read(), pos++ ) {
+        assertEquals("Files differed at position " + pos,  a, b);
+      }
+      assertTrue("Files not the same length.", a == -1 && a == b);
+    }
+    finally {
+      expected.close();;
+      actual.close();
+    }
```
[the remark this answers is not in the document]
Comment (m4dc4p) [reply androidtest/src/androidTest/java/com/tozny/e3db/AndroidCryptoTest.java:400]: Done!

**Review by emanb29 — APPROVED** [review]
The new function names are definitely more clear.

**Pull request merged**